File: src/main/java/com/github/tanokun/tanorpg/player/quest/QuestManager.java
package com.github.tanokun.tanorpg.player.quest;

import com.github.tanokun.tanorpg.TanoRPG;
import com.github.tanokun.tanorpg.game.item.ItemRarityType;
import com.github.tanokun.tanorpg.player.quest.action.Action;
import com.github.tanokun.tanorpg.player.quest.action.ActionType;
import com.github.tanokun.tanorpg.player.quest.condition.Condition;
import com.github.tanokun.tanorpg.player.quest.condition.ConditionType;
import com.github.tanokun.tanorpg.player.quest.task.Task;
import com.github.tanokun.tanorpg.player.quest.task.TaskType;
import com.github.tanokun.tanorpg.util.command.CommandContext;
import com.github.tanokun.tanorpg.util.io.Config;
import com.github.tanokun.tanorpg.util.io.Folder;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class QuestManager {
    public static final String PX = "§6[§a-｜ §b§lMission§a ｜-§6] §7=> §a";

    HashMap<Integer, ArrayList<Quest>> quests = new HashMap<>();

    public QuestManager(Player p, Plugin plugin){
        if (!TanoRPG.getPlugin().getDataManager().isInitFile()) new Config("quests" + File.separator + "Sample.yml", plugin).saveDefaultConfig();
        if (p != null) p.sendMessage(TanoRPG.PX + "§bLoading quest configs...");
        else Bukkit.getConsoleSender().sendMessage("[TanoRPG] §bLoading quest configs...");
        loadQuest(p);
        if (p != null) p.sendMessage(TanoRPG.PX + " ");
        else Bukkit.getConsoleSender().sendMessage(" ");
    }

    public void loadQuest(Player p){
        String path = "";
        String filePath = "";
        HashSet<String> errors = new HashSet<>();
        errors.add("§a    Quest configs loaded without errors.");

        CommandContext cc = new CommandContext(null, new String[]{""});

        try {
            path = "quests";
            for (Config config : new Folder(path, TanoRPG.getPlugin()).getFiles()) {
                filePath = path + File.separator + config.getName() + File.separator;

                path = "options.name";
                String name = config.getConfig().getString(path, "unknown");

                path = "options.npc";
                int npcId = config.getConfig().getInt(path, 0);

                path = "options.difficulty";
                ItemRarityType difficulty = ItemRarityType.valueOf(config.getConfig().getString(path, "COMMON"));

                path = "options.lore";
                List<String> lore = new ArrayList<>();
                config.getConfig().getStringList(path).stream().forEach(t -> lore.add("§f" + t));

                path = "options.result";
                List<String> result = new ArrayList<>();
                config.getConfig().getStringList(path).stream().forEach(t -> result.add("§f - " + t));

                path = "conditions";
                List<Condition> conditions = new ArrayList<>();
                for (String condition : config.getConfig().getConfigurationSection(path).getKeys(false)){
                    conditions.add(ConditionType.valueOf(condition).getCondition().getConstructor(Config.class).newInstance(config));
                }

                path = "tasks";
                List<Task> tasks = new ArrayList<>();
                for (String task : config.getConfig().getStringList(path)){
                    cc.init(null, task.split(" "));
                    tasks.add(TaskType.valueOf(cc.args.remove(0)).getTask().getConstructor(CommandContext.class).newInstance(cc));
                }

                path = "showQuestActions";
                List<Action> showQuestActions = new ArrayList<>();
                for (String showQuestAction : config.getConfig().getStringList(path)){
                    cc.init(null, showQuestAction.split(" "));
                    showQuestActions.add(ActionType.valueOf(cc.args.remove(0)).getAction().getConstructor(CommandContext.class).newInstance(cc));
                }

                path = "startQuestActions";
                List<Action> startQuestActions = new ArrayList<>();
                for (String startQuestAction : config.getConfig().getStringList(path)){
                    cc.init(null, startQuestAction.split(" "));
                    startQuestActions.add(ActionType.valueOf(cc.args.remove(0)).getAction().getConstructor(CommandContext.class).newInstance(cc));
                }

                path = "finishQuestActions";
                List<Action> finishQuestActions = new ArrayList<>();
                for (String finishQuestAction : config.getConfig().getStringList(path)){
                    cc.init(null, finishQuestAction.split(" "));
                    finishQuestActions.add(ActionType.valueOf(cc.args.remove(0)).getAction().getConstructor(CommandContext.class).newInstance(cc));
                }

                path = "cancelQuestActions";
                List<Action> cancelQuestActions = new ArrayList<>();
                for (String cancelQuestAction : config.getConfig().getStringList(path)){
                    cc.init(null, cancelQuestAction.split(" "));
                    cancelQuestActions.add(ActionType.valueOf(cc.args.remove(0)).getAction().getConstructor(CommandContext.class).newInstance(cc));
                }

                ArrayList<Quest> quests = this.quests.get(npcId) == null ? new ArrayList<>() : this.quests.get(npcId);
                quests.add(new Quest(name, npcId, lore, result, difficulty, conditions, tasks, showQuestActions, startQuestActions, finishQuestActions, cancelQuestActions));
                this.quests.put(npcId, quests);

            }
        }catch (Exception e){
            errors.remove("§a    Quest configs loaded without errors.");
            errors.add("§c    " + e.getMessage() + "§7" + "(Path: " + filePath + path + ")");
        }

        showErrors(errors, p);
    }

    private void showErrors(HashSet<String> errors, Player p){
        if (p != null) errors.stream().forEach(e -> p.sendMessage( e));
        else errors.stream().forEach(e -> Bukkit.getConsoleSender().sendMessage(e));
    }

    public Quest getQuest(int npcId, String name){
        if (quests.get(npcId) == null) return null;
        final Quest[] r = {null};
        quests.get(npcId).stream()
                .filter(q -> q.getName().equals(name))
                .forEach(q -> r[0] = q);
        return r[0];
    }

    public List<Quest> getQuests(int npcId){
        return quests.get(npcId);
    }
}

File: src/main/java/com/github/tanokun/tanorpg/player/quests/QuestManager.kt
package com.github.tanokun.tanorpg.player.quests

import com.github.tanokun.tanorpg.TanoRPG
import com.github.tanokun.tanorpg.TanoRPG.Companion.plugin
import com.github.tanokun.tanorpg.game.item.ItemRarityType
import com.github.tanokun.tanorpg.player.quests.actions.QuestAction
import com.github.tanokun.tanorpg.player.quests.conditions.QuestCondition
import com.github.tanokun.tanorpg.player.quests.tasks.QuestTask
import com.github.tanokun.tanorpg.util.io.Config
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.io.File
import java.util.function.Consumer


class QuestManager(p: Player?) {
    companion object {
        const val PX = "§6[§a-｜ §b§lQuest§a ｜-§6] §7=> "
    }

    private var quests = HashMap<Int, ArrayList<QuestModel>>()

    private var questsByID = HashMap<String, QuestModel>()

    init {
        val errors = HashSet<String>()
        if (p != null) p.sendMessage(TanoRPG.PX + "§bLoading quest configs...") else Bukkit.getConsoleSender()
            .sendMessage("[TanoRPG] §bLoading quest configs...")
        loadQuest(Config(plugin, "quests").file, errors)
        showErrors(errors, p)
        if (p != null) p.sendMessage(TanoRPG.PX + " ") else Bukkit.getConsoleSender().sendMessage(" ")
    }

    private fun loadQuest(tempFile: File, errors: HashSet<String>) {
        var path = "quests"
        var filePath = ""
        val files = tempFile.listFiles()
        for (tmpFile in files ?: arrayOfNulls(0)) {
            if (tmpFile.isDirectory) {
                loadQuest(tmpFile, errors)
            } else {
                try {
                    val config = Config(plugin, tmpFile, tmpFile.name)
                    filePath = "..." + File.separator + tempFile.name + File.separator + config.fileName + File.separator
                    path = "options.name"
                    val name = config.config.getString(path, "unknown")!!

                    path = "options.id"
                    val id = config.config.getString(path, "unknown")!!

                    path = "options.npc"
                    val npcId = config.config.getInt(path, 0)

                    path = "options.difficulty"
                    val difficulty = ItemRarityType.valueOf(config.config.getString(path, "COMMON")!!)

                    path = "options.lore"
                    val lore: MutableList<String> = ArrayList()
                    config.config.getStringList(path).forEach(Consumer { t: String -> lore.add("§f$t")})

                    path = "options.result"
                    val result: MutableList<String> = ArrayList()
                    config.config.getStringList(path).forEach(Consumer { t: String -> result.add("§f * $t") })

                    path = "options.flags"
                    val flags: MutableList<String> = config.config.getStringList(path)

                    /*
                    val cantToDoQuestShow = config.config.getBoolean("options.cantToDoQuestShow", false)
                    val minutes = config.config.getLong("options.minutes", 0)

                    path = "conditions"
                    val conditions: MutableList<com.github.tanokun.tanorpg.player.quest.condition.Condition> =
                        ArrayList<com.github.tanokun.tanorpg.player.quest.condition.Condition>()
                    for (condition in config.config.getConfigurationSection(path)!!.getKeys(false)) {
                        conditions.add(
                            ConditionType.valueOf(condition!!).condition.getConstructor(Config::class.java)
                                .newInstance(config)
                        )
                    }

                     */

                    path = "conditions"
                    val conditions = ArrayList<QuestCondition>()
                    config.config.getConfigurationSection(path)?.let {
                        it.getKeys(false).forEach { id ->
                            conditions.add(QuestCondition.getCondition(id, config))
                        }
                    }


                    path = "showQuestActions"
                    val showQuestActions: MutableList<QuestAction> = ArrayList()
                    config.config.getStringList(path).forEach { raw: String ->
                        val type = raw.split(" ")[0]
                        showQuestActions.add(QuestAction.getAction(type, raw.replace("$type ", "")))
                    }

                    path = "finishQuestActions"
                    val finishQuestActions: MutableList<QuestAction> = ArrayList()
                    config.config.getStringList(path).forEach { raw: String ->
                        val type = raw.split(" ")[0]
                        finishQuestActions.add(QuestAction.getAction(type, raw.replace("$type ", "")))
                    }

                    path = "cancelQuestActions"
                    val cancelQuestActions: MutableList<QuestAction> = ArrayList()
                    config.config.getStringList(path).forEach { raw: String ->
                        val type = raw.split(" ")[0]
                        cancelQuestActions.add(QuestAction.getAction(type, raw.replace("$type ", "")))
                    }

                    val questIndex: MutableList<QuestIndex> = ArrayList()
                    for (key in config.config.getKeys(false)) {
                        if (!key.contains("index_")) continue
                        path = "$key.actions"
                        val indexAction: MutableList<QuestAction> = ArrayList()
                        for (S_Index in config.config.getStringList(path)) {
                            val type = S_Index.split(" ")[0]
                            indexAction.add(QuestAction.getAction(type, S_Index.replace("$type ", "")))
                        }

                        path = "$key.tasks"
                        val indexTasks: MutableList<QuestTask<Any>> = ArrayList()
                        for (raw in config.config.getStringList(path)) {
                            val type = raw.split(" ")[0]
                            val task: QuestTask<Any> = QuestTask.getTask(type, raw.replace("$type ", ""), config)
                            indexTasks.add(task)
                        }
                        questIndex.add(QuestIndex(indexTasks, indexAction))
                    }
                    val quests = if (quests[npcId] == null) ArrayList() else quests[npcId]!!
                    val quest = QuestModel(id, name, npcId, lore, result, difficulty, conditions, showQuestActions, cancelQuestActions, finishQuestActions, questIndex, flags)
                    quests.add(quest)
                    this.quests[npcId] = quests
                    this.questsByID[id] = quest
                } catch (e: Exception) {
                    errors.add("§c    " + e.message + "§7" + "(Path: " + filePath + path + ")")
                }
            }
        }
    }

    private fun showErrors(errors: HashSet<String>, p: Player?) {
        if (errors.size == 0) errors.add("§a    Quest configs loaded without errors.")
        if (p != null) errors.forEach(Consumer { string: String -> p.sendMessage(string) })
        else errors.forEach { e: String -> Bukkit.getConsoleSender().sendMessage(e) }
    }

    fun getQuests(npc: Int): ArrayList<QuestModel>? = quests[npc]

    fun getQuest(id: String): QuestModel? = questsByID[id]

}